### PR TITLE
Custom folders part 10

### PR DESCRIPTION
**Pull request by Screenfeed**
Bulk page.

### CODE REVIEW

**Comment by Tabrisrp** [comment]
should I review each commit?

**Comment by Screenfeed** [comment]
As you want. The "merge branch" commit is small this time.
But it's a big branch indeed.

**Comment by Screenfeed** [comment]
Poke.

File: inc/classes/class-imagify-files-list-table.php [hunk 1]
```diff
@@ -141,6 +140,23 @@ public function prepare_items() {
...
+		if ( $status_filter ) {
+			// Display files optimized, not optimized, or with error.
+			$where .= $where ? ' AND ' : 'WHERE ';
+
+			switch ( $status_filter ) {
+				case 'optimized':
+					$where .= "( status = 'success' OR status = 'already_optimized' )";
+					break;
+				case 'unoptimized':
+					$where .= 'status IS NUUL';
```
[the remark this answers is not in the document]
Comment (Screenfeed) [reply inc/classes/class-imagify-files-list-table.php]: Yes: https://github.com/wp-media/imagify-plugin/commit/ae866a456f338166b6f24ff5b81395140de1b712

File: inc/classes/class-imagify-views.php [hunk 2]
```diff
@@ -319,6 +454,7 @@ public function get_files_page_slug() {
 	 * @return string|bool      The page contents. False if the template doesn't exist.
 	 */
 	public function get_template( $template, $data = array() ) {
+		$path = str_replace( '_', '-', $template );
 		$path = IMAGIFY_INC_PATH . 'views/' . $template . '.php';
 
 		if ( ! imagify_get_filesystem()->exists( $path ) ) {
```
[the remark this answers is not in the document]
Comment (Screenfeed) [reply inc/classes/class-imagify-views.php]: Why wouldn't it be? The templates are inside the plugin, so if the class file is readable, the template should be too, no?

File: inc/classes/class-imagify-admin-ajax-post.php [hunk 3]
```diff
@@ -261,41 +264,54 @@ public function imagify_bulk_upload_callback() {
...
-			$data['success'] = false;
-			$data['error']   = $fullsize_data['error'];
+			$data['success']    = false;
+			$data['error_code'] = '';
+			$data['error']      = isset( $fullsize_data['error'] ) ? (string) $fullsize_data['error'] : '';
+
+			if ( ! $attachment->has_error() ) {
+				$data['error_code'] = 'already-optimized';
+			} else {
+				$message = imagify_translate_api_message( 'You\'ve consumed all your data. You have to upgrade your account to continue.' );
```
[the remark this answers is not in the document]
Comment (Screenfeed) [reply inc/classes/class-imagify-admin-ajax-post.php]: Not here nor the next one. This is a message from the API: the plugin gets this message from the API answer, then this function improves the message and translate it.
Here I feed the function with a message that the API could have returned, to get the real translated message.
See: https://github.com/wp-media/imagify-plugin/blob/master/inc/functions/api.php#L287.

File: inc/3rd-party/nextgen-gallery/inc/classes/class-imagify-ngg-attachment.php [hunk 4]
```diff
@@ -229,8 +275,9 @@ public function update_metadata_size() {
 			$metadata['height']         = $size[1];
 			$metadata['full']['width']  = $size[0];
 			$metadata['full']['height'] = $size[1];
+			$this->image->meta_data     = $metadata;
```
[the remark this answers is not in the document]
Comment (Screenfeed) [reply inc/3rd-party/nextgen-gallery/inc/classes/class-imagify-ngg-attachment.php]: Indeed. Fixed: https://github.com/wp-media/imagify-plugin/commit/f2e7f386debb9ae4a6735263dce0120af6bd3ada.

**Pull request merged**